### PR TITLE
lag interfaces resource module

### PR DESCRIPTION
**Pull request by rohitthakur2590**
Signed-off-by: rohitthakur2590 <rohitthakur2590@outlook.com>
##### SUMMARY
- Add new resource module vyos_lag_interfaces.
- Targets model - ansible-network/resource_module_models#25
[vyos_lag_interfaces]

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
vyos_lag_interfaces

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Review by pabelanger — COMMENTED** [review]
Can we get away from using eth3 for testing?

**Comment by rohitthakur2590** [comment]
recheck

File: test/integration/targets/vyos_lag_interfaces/tests/cli/merged.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+          - name: bond1
+            hash-policy: "layer2+3"
+            mode: "active-backup"
+            members:
+                - member: eth2
+            primary: eth2
+      state: merged
+  register: result
+
+- name: Assert that before dicts were correctly generated
```
Comment (pabelanger) [inline test/integration/targets/vyos_lag_interfaces/tests/cli/merged.yaml]: This appears to be failing, but I don't really understand the logic here

Comment (rohitthakur2590) [reply test/integration/targets/vyos_lag_interfaces/tests/cli/merged.yaml]: We are comparing that config before  actual merged state  operation and config shown in output result ( 'before[]') after executing merge operation should be same.

Comment (pabelanger) [reply test/integration/targets/vyos_lag_interfaces/tests/cli/merged.yaml]: I suspect the before config is in the wrong state before we are running our test. It is hard to see what the issue is, but we should be able to debug this.

Comment (rohitthakur2590) [reply test/integration/targets/vyos_lag_interfaces/tests/cli/merged.yaml]: Yeah seems like some sort of race condition , or delay . I am gonna try to explicitly set the state giving enough time for system to get ready for next state operation  and  yeah we  should be able to debug this.

**Review by NilashishC — CHANGES_REQUESTED** [review]
We should be using `block` and `always` in test files to make sure that we cleanup even if an assertion fails.

File: lib/ansible/module_utils/network/vyos/facts/lag_interfaces/lag_interfaces.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                obj['name'] = lag.strip("'")
+                if members:
+                    obj['members'] = members
+
+                if obj:
+                    objs.append(obj)
+
+        facts = {}
+        if objs:
+            facts['lag_interfaces'] = objs
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/vyos/facts/lag_interfaces/lag_interfaces.py]: `validate_config` call is missing, IMO that should be added. Please refer to [this](https://github.com/ansible/network/pull/42/files#diff-40a2b0039c14ce8fc94b0244e9ce8614R61).

File: lib/ansible/module_utils/network/vyos/utils/utils.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    if have_list and not want_list:
+        diff = None
+    elif not have_list:
+        diff = want_list
+    else:
+        diff = [i for i in have_list + want_list if i in want_list and i not in have_list]
+    return diff
+
+
+def add_bond_members(want_item, have_item):
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/vyos/utils/utils.py]: These seem to be resource specific and not generic utility methods that can be reused by modules apart from lag_interfaces.

File: lib/ansible/modules/network/vyos/vyos_facts.py [hunk 4]
```diff
@@ -1,65 +1,88 @@
...
-    is running VyOS.  This module prepends all of the
-    base network fact keys with U(ansible_net_<fact>).  The facts
-    module will always collect a base set of facts from the device
-    and can enable or disable collection of additional facts.
+  - Collects facts from network devices running the vyos operating
+    system. This module places the facts gathered in the fact tree keyed by the
+    respective resource name.  The facts module will always collect a
+    base set of facts from the device and can enable or disable
+    collection of additional facts.
+author: Rohit Thakur (@rohitthakur2590)
```
Comment (NilashishC) [inline lib/ansible/modules/network/vyos/vyos_facts.py]: The `author` key is a list and can include the previous author as well.

File: lib/ansible/modules/network/vyos/vyos_facts.py [hunk 5]
```diff
@@ -1,65 +1,88 @@
...
+    description:
+      - When supplied, this argument will restrict the facts collected
+        to a given subset. Possible values for this argument include
+        all and the resources like lag interfaces.
+        Can specify a list of values to include a larger subset. Values
+        can also be used with an initial C(M(!)) to specify that a
+        specific subset should not be collected.
+    required: false
+    version_added: "2.9"
+    choices: ['all', 'lag_interfaces']
```
Comment (NilashishC) [inline lib/ansible/modules/network/vyos/vyos_facts.py]: `!{{ resource }}` can also be a choice.

File: lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+                commands.extend(self._state_merged(want_lag=item, have_lag=obj_in_have))
+        elif state == 'replaced':
+            for item in want:
+                name = item['name']
+                obj_in_have = search_obj_in_list(name, have)
+                commands.extend(self._state_replaced(want_lag=item, have_lag=obj_in_have))
+        return commands
+
+    @staticmethod
+    def _state_replaced(**kwargs):
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py]: Using kwargs is probably redundant cause we will always be passing the same arguments to the _state_* methods.

File: lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+        commands.extend(
+            Lag_interfaces._state_merged(
+                want_lag=want_lag,
+                have_lag=have_lag
+            )
+        )
+        return commands
+
+    @staticmethod
+    def _state_overridden(**kwargs):
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py]: Same as above.

File: lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+            commands.extend(
+                Lag_interfaces._state_replaced(
+                    want_lag=lag,
+                    have_lag=lag_in_have
+                )
+            )
+        return commands
+
+    @staticmethod
+    def _state_merged(**kwargs):
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py]: Same as above.

File: lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+                Lag_interfaces._render_set_commands(
+                    want_element={
+                        'lag': want_lag
+                    }
+                )
+            )
+        return commands
+
+    @staticmethod
+    def _state_deleted(**kwargs):
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py]: Same as above.

File: lib/ansible/modules/network/vyos/vyos_lag_interfaces.py [hunk 10]
```diff
@@ -0,0 +1,553 @@
...
+  returned: when changed
+  type: list
+  sample: >
+    The configuration returned will always be in the same format
+     of the parameters above.
+commands:
+  description: The set of commands pushed to the remote device.
+  returned: always
+  type: list
+  sample: ['command 1', 'command 2', 'command 3']
```
Comment (NilashishC) [inline lib/ansible/modules/network/vyos/vyos_lag_interfaces.py]: A couple of sample commands here would be good.

**Comment by rohitthakur2590** [comment]
recheck

**Comment by rohitthakur2590** [comment]
recheck

File: lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+            commands.extend(self._purge_attribs(have))
+        return commands
+
+    def _render_updates(self, want, have):
+        commands = []
+
+        try:
+            temp_have_members = have.pop('members', None)
+            temp_want_members = want.pop('members', None)
+        except BaseException:
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py]: We would not be hitting an exception as we're using pop() with a default value.

File: lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+            diff = want_arp_target
+        else:
+            have_arp_monitor = have_list.get(dict_name) or {}
+            if have_arp_monitor and lst in have_arp_monitor:
+                have_arp_target = have_arp_monitor[lst]
+
+            diff = list_diff_want_only(want_arp_target, have_arp_target)
+        return diff
+
+    def _compute_command(self, name, attrib, value=None, set_flag=True, type='bonding'):
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py]: `type` can be determined from name of the interface - you can refer to `get_interface_type` which is already added in utils.

File: lib/ansible/modules/network/vyos/vyos_lag_interfaces.py [hunk 13]
```diff
@@ -0,0 +1,555 @@
...
+    'status': ['preview'],
+    'supported_by': 'network'
+}
+
+DOCUMENTATION = """
+---
+module: vyos_lag_interfaces
+version_added: 2.9
+short_description: Manages attributes of link aggregation groups on VyOS network devices.
+description: This module manages attributes of link aggregation groups on VyOS network devices.
```
Comment (NilashishC) [inline lib/ansible/modules/network/vyos/vyos_lag_interfaces.py:43]: Please add the `notes` section as discussed in the other PR.

File: lib/ansible/modules/network/vyos/vyos_lag_interfaces.py [hunk 14]
```diff
@@ -0,0 +1,555 @@
...
+    lag_interfaces import Lag_interfacesArgs
+from ansible.module_utils.network.vyos.config.lag_interfaces.lag_interfaces import Lag_interfaces
+
+
+def main():
+    """
+    Main entry point for module execution
+
+    :returns: the result form module invocation
+    """
```
Comment (NilashishC) [inline lib/ansible/modules/network/vyos/vyos_lag_interfaces.py:549]: Needs a `required_if` as discussed in the other PR.

File: lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+                )
+        return commands
+
+    def _add_bond_members(self, want, have):
+        commands = []
+        diff_members = get_lst_diff_for_dicts(want, have, 'members')
+        if diff_members:
+            for key in diff_members:
+                commands.append(
+                    self._compute_command(key['member'], 'bond-group', want['name'], type='ethernet')
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py:278]: `type` can be determined from the name of the interface.

File: lib/ansible/modules/network/vyos/vyos_lag_interfaces.py (hunk 13 above)
Comment (rohitthakur2590) [reply lib/ansible/modules/network/vyos/vyos_lag_interfaces.py:43]: Done

File: lib/ansible/modules/network/vyos/vyos_lag_interfaces.py (hunk 14 above)
Comment (rohitthakur2590) [reply lib/ansible/modules/network/vyos/vyos_lag_interfaces.py:549]: Done

File: lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py (hunk 15 above)
Comment (rohitthakur2590) [reply lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py:278]: As discussed logic is little bit different here and only needed to pass interface in two cases .though I have renamed the variables to somewhat align with  vyos interfaces compute function .

File: lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py (hunk 12 above)
Comment (rohitthakur2590) [reply lib/ansible/module_utils/network/vyos/config/lag_interfaces/lag_interfaces.py]: Same

**Review by NilashishC — APPROVED** [review]
+1 Please also add prepare_vyos_tests as a dependency for the integration test role.

**Pull request merged**